File: examples/integrations/hr-workforce-itrent/push_workforce.cs
// Push a weekly workforce snapshot to the Ingest service from MHR iTrent (minimal C# example).
//
// This is a .NET 10 "file-based app": run it directly with `dotnet run push_workforce.cs`
// - no .csproj, no build step. It GETs a weekly summary from iTrent's OData API,
// reads the few columns it needs, maps them to the `weekly_workforce` schema, and
// POSTs one submission. Only the BCL is used.
//
// The request JSON is built by hand (file-based apps disable reflection-based
// serialization by default), which also makes the payload shape explicit. The
// reflection-free JsonDocument is used to read both the iTrent and Ingest responses.
//
// Usage:
//   set INGEST_BASE_URL=https://ingest.example.org
//   set INGEST_API_KEY=abc12345.your-secret-here
//   dotnet run push_workforce.cs -- [--source-url URL] [--dry-run]

using System.Globalization;
using System.Text;
using System.Text.Json;

const string SchemaName = "weekly_workforce";
const string DefaultSourceUrl = "http://localhost:8000/sample_response.json";

bool dryRun = args.Contains("--dry-run");

string sourceUrl = DefaultSourceUrl;
int urlFlag = Array.IndexOf(args, "--source-url");
if (urlFlag >= 0 && urlFlag + 1 < args.Length)
    sourceUrl = args[urlFlag + 1];

// 1. Query iTrent for just the columns we need. In production the URL carries the
//    OData query, e.g.
//    .../odata/v1/WeeklyWorkforceSummary?$select=activeEmployees,absenceSickness,contingentWorkers,overtimeHours&$filter=organisationUnit eq 'Waste Services'
JsonElement row = await FetchWeek(sourceUrl);

// Weekly cadence: one sample per week bucket. Use the week-ending date (UTC).
string timestamp = $"{row.GetProperty("weekEnding").GetString()}T00:00:00Z";

// 2. Map the iTrent columns -> schema values.
var samples = new List<string>
{
    Sample("employees_active", Num(row.GetProperty("activeEmployees").GetInt32())),
    Sample("sick_leave", Num(row.GetProperty("absenceSickness").GetInt32())),
    Sample("contractors", Num(row.GetProperty("contingentWorkers").GetInt32())),
};

// overtime_hours is optional: only send when there was any overtime to report.
double overtime = row.TryGetProperty("overtimeHours", out var ot) ? ot.GetDouble() : 0;
if (overtime > 0)
    samples.Add(Sample("overtime_hours", Num(overtime)));

string body = "{\"samples\":[" + string.Join(",", samples) + "]}";

if (dryRun)
{
    Console.WriteLine(Pretty(body));
    return;
}

string? baseUrl = Environment.GetEnvironmentVariable("INGEST_BASE_URL");
string? apiKey = Environment.GetEnvironmentVariable("INGEST_API_KEY");
if (string.IsNullOrWhiteSpace(baseUrl) || string.IsNullOrWhiteSpace(apiKey))
{
    Console.Error.WriteLine("Set INGEST_BASE_URL and INGEST_API_KEY (or use --dry-run).");
    Environment.Exit(1);
}

await PostSubmission(baseUrl!.TrimEnd('/'), apiKey!, body);

// GET the OData feed and return the first (and only) matching team-week row.
static async Task<JsonElement> FetchWeek(string url)
{
    using var http = new HttpClient();
    string payload = await http.GetStringAsync(url);
    using var doc = JsonDocument.Parse(payload);
    var rows = doc.RootElement.GetProperty("value");
    if (rows.GetArrayLength() == 0)
    {
        Console.Error.WriteLine("iTrent returned no rows for this week - nothing to submit.");
        Environment.Exit(1);
    }
    // Clone so the value survives the JsonDocument being disposed.
    return rows[0].Clone();
}

// One sample object as a JSON string. `valueJson` is already JSON-encoded.
string Sample(string valueName, string valueJson) =>
    $"{{\"schemaName\":{Str(SchemaName)},\"valueName\":{Str(valueName)},\"value\":{valueJson},\"timestamp\":{Str(timestamp)},\"note\":null}}";

static string Num(double n) => n.ToString(CultureInfo.InvariantCulture);

static string Str(string s)
{
    var sb = new StringBuilder("\"");
    foreach (char ch in s)
        sb.Append(ch switch
        {
            '\\' => "\\\\",
            '"' => "\\\"",
            '\n' => "\\n",
            '\r' => "\\r",
            '\t' => "\\t",
            _ => ch.ToString(),
        });
    return sb.Append('"').ToString();
}

static string Pretty(string json)
{
    using var doc = JsonDocument.Parse(json);
    using var stream = new MemoryStream();
    using (var writer = new Utf8JsonWriter(stream, new JsonWriterOptions { Indented = true }))
        doc.RootElement.WriteTo(writer);
    return Encoding.UTF8.GetString(stream.ToArray());
}

static async Task PostSubmission(string baseUrl, string apiKey, string body)
{
    using var http = new HttpClient();
    using var content = new StringContent(body, Encoding.UTF8, "application/json");
    content.Headers.Add("X-Api-Key", apiKey);

    using var response = await http.PostAsync($"{baseUrl}/api/submissions", content);
    string payload = await response.Content.ReadAsStringAsync();

    if (response.IsSuccessStatusCode)
    {
        using var doc = JsonDocument.Parse(payload);
        Console.WriteLine($"Created submission {doc.RootElement.GetProperty("id").GetString()}");
        if (doc.RootElement.TryGetProperty("warnings", out var warnings))
            foreach (var w in warnings.EnumerateArray())
                Console.WriteLine($"  warning: {w.GetString()}");
        return;
    }

    Console.Error.WriteLine($"Submission failed: HTTP {(int)response.StatusCode}");
    try
    {
        using var doc = JsonDocument.Parse(payload);
        if (doc.RootElement.TryGetProperty("errors", out var errors) && errors.ValueKind == JsonValueKind.Array)
            foreach (var e in errors.EnumerateArray())
                Console.Error.WriteLine($"  error: {e.GetString()}");
        else if (doc.RootElement.TryGetProperty("detail", out var detail))
            Console.Error.WriteLine($"  {detail.GetString()}");
    }
    catch (JsonException)
    {
        Console.Error.WriteLine($"  {payload}");
    }
    Environment.Exit(1);
}

File: examples/integrations/hr-workforce-itrent/push_workforce.py
#!/usr/bin/env python3
"""Pull a weekly workforce summary from MHR iTrent's OData API and push it to Ingest.

A MINIMAL, dependency-free example of the "vendor API" integration style for HR
data, aimed at MHR iTrent. iTrent exposes OData feeds, so you query just the few
columns you need with $select, map them to the `weekly_workforce` schema, and POST
one weekly submission.

For a self-contained run, the default source is the bundled sample served by
`python -m http.server` (see README). Point --source-url at the real iTrent OData
endpoint in production.

Usage:
    set INGEST_BASE_URL=https://ingest.example.org
    set INGEST_API_KEY=abc12345.your-secret-here
    python push_workforce.py [--source-url URL] [--dry-run]
"""
from __future__ import annotations

import argparse
import json
import os
import sys
import urllib.error
import urllib.request

SCHEMA_NAME = "weekly_workforce"
DEFAULT_SOURCE_URL = "http://localhost:8000/sample_response.json"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Submit a weekly workforce snapshot to Ingest from iTrent.")
    parser.add_argument(
        "--source-url",
        default=os.environ.get("SOURCE_URL", DEFAULT_SOURCE_URL),
        help="iTrent OData URL returning the weekly summary JSON.",
    )
    parser.add_argument(
        "--dry-run",
        action="store_true",
        help="Build and print the payload but do not call the Ingest API.",
    )
    return parser.parse_args()


def fetch_week(url: str) -> dict[str, object]:
    # In production the URL carries the OData query, e.g.
    # .../odata/v1/WeeklyWorkforceSummary?$select=activeEmployees,absenceSickness,contingentWorkers,overtimeHours&$filter=organisationUnit eq 'Waste Services'
    request = urllib.request.Request(url, method="GET")
    with urllib.request.urlopen(request) as response:
        payload = json.loads(response.read().decode("utf-8"))
    # OData returns matching rows under "value"; one team-week is one row.
    rows = payload.get("value") or []
    if not rows:
        raise SystemExit("iTrent returned no rows for this week - nothing to submit.")
    return rows[0]


def build_samples(row: dict[str, object]) -> list[dict[str, object]]:
    # Weekly cadence: one sample per week bucket, keyed on the week-ending date.
    timestamp = f"{row['weekEnding']}T00:00:00Z"

    def sample(value_name: str, value: object) -> dict[str, object]:
        return {
            "schemaName": SCHEMA_NAME,
            "valueName": value_name,
            "value": value,
            "timestamp": timestamp,
            "note": None,
        }

    # Map the iTrent columns -> schema values.
    samples = [
        sample("employees_active", int(row["activeEmployees"])),
        sample("sick_leave", int(row["absenceSickness"])),
        sample("contractors", int(row["contingentWorkers"])),
    ]

    # overtime_hours is optional: only send when there was overtime to report.
    overtime = float(row.get("overtimeHours") or 0)
    if overtime > 0:
        samples.append(sample("overtime_hours", overtime))

    return samples


def post_submission(base_url: str, api_key: str, body: dict[str, object]) -> None:
    url = base_url.rstrip("/") + "/api/submissions"
    data = json.dumps(body).encode("utf-8")
    request = urllib.request.Request(
        url,
        data=data,
        method="POST",
        headers={"Content-Type": "application/json", "X-Api-Key": api_key},
    )
    try:
        with urllib.request.urlopen(request) as response:
            payload = json.loads(response.read().decode("utf-8"))
            print(f"Created submission {payload['id']}")
            for warning in payload.get("warnings", []):
                print(f"  warning: {warning}")
    except urllib.error.HTTPError as error:
        detail = error.read().decode("utf-8")
        print(f"Submission failed: HTTP {error.code}", file=sys.stderr)
        try:
            problem = json.loads(detail)
            for err in problem.get("errors", []):
                print(f"  error: {err}", file=sys.stderr)
            if not problem.get("errors"):
                print(f"  {problem.get('detail', detail)}", file=sys.stderr)
        except json.JSONDecodeError:
            print(f"  {detail}", file=sys.stderr)
        raise SystemExit(1)


def main() -> None:
    args = parse_args()
    row = fetch_week(args.source_url)
    body = {"samples": build_samples(row)}

    if args.dry_run:
        print(json.dumps(body, indent=2))
        return

    base_url = os.environ.get("INGEST_BASE_URL")
    api_key = os.environ.get("INGEST_API_KEY")
    if not base_url or not api_key:
        raise SystemExit("Set INGEST_BASE_URL and INGEST_API_KEY (or use --dry-run).")

    post_submission(base_url, api_key, body)


if __name__ == "__main__":
    main()
